core/geometry/src/geometry3d/YawPitchRollAngles.ts

```diff
@@ -158,53 +218,113 @@ export class YawPitchRollAngles {
   public static tryFromTransform(transform: Transform): {
     origin: Point3d;
     angles: YawPitchRollAngles | undefined;
   } {
-    // bundle up the transform's origin with the angle data extracted from the transform
     return {
-      angles: YawPitchRollAngles.createFromMatrix3d(transform.matrix),
       origin: Point3d.createFrom(transform.origin),
+      angles: YawPitchRollAngles.createFromMatrix3d(transform.matrix),
     };
   }
-  /** Attempts to create a YawPitchRollAngles object from an Matrix3d
-   * * This conversion fails if the matrix is not rigid (unit rows and columns, transpose is inverse)
+  /** Attempts to create a YawPitchRollAngles object from a Matrix3d
+   * * This conversion fails if the matrix is not rigid (unit rows and columns, and transpose is inverse)
    * * In the failure case the method's return value is `undefined`.
-   * * In the failure case, if the optional result was supplied, that result will nonetheless be filled with a set of angles.
+   * * In the failure case, if the optional result was supplied, that result will nonetheless be filled with
+   * a set of angles.
    */
   public static createFromMatrix3d(matrix: Matrix3d, result?: YawPitchRollAngles): YawPitchRollAngles | undefined {
-    const s1 = matrix.at(2, 0);
-    const c1 = Math.sqrt(matrix.at(2, 1) * matrix.at(2, 1) + matrix.at(2, 2) * matrix.at(2, 2));
-    const pitchA = Angle.createAtan2(s1, c1); // with positive cosine
-    const pitchB = Angle.createAtan2(s1, -c1); // with negative cosine
-    const angles = result ? result : new YawPitchRollAngles(); // default undefined . . .
-    if (c1 < Geometry.smallAngleRadians) { // This is a radians test !!!
+    /**
+     * The rotation matrix for is
+     *
+     * Matrix3d.createRowValues(
+     *      cz * cy, -(sz * cx + cz * sy * sx), (sz * sx - cz * sy * cx),
+     *      sz * cy, (cz * cx - sz * sy * sx), -(cz * sx + sz * sy * cx),
+     *      sy, cy * sx, cy * cx
+     * );
+     *
+     * where cx = cos(x), sx = sin(x), cy = cos(y), sy = sin(y), cz = cos(z), and sz = sin(z)
+     */
+    const sy = matrix.at(2, 0); // sin(y)
+    const cy = Math.sqrt(matrix.at(2, 1) * matrix.at(2, 1) + matrix.at(2, 2) * matrix.at(2, 2)); // |cos(y)|
+    const pitchA = Angle.createAtan2(sy, cy); // with positive cosine
+    const pitchB = Angle.createAtan2(sy, -cy); // with negative cosine
+    const angles = result ? result : new YawPitchRollAngles();
+    /**
+     * If cos(y) = 0 then y = +-90 degrees so we have a gimbal lock.
+     * This means x and z can be anything as long as their sum x + z is constant.
+     * so we can pick z = 0 and calculate x (or pick x = 0 and calculate z).
+     * Math details can be found
+     * https://en.wikipedia.org/wiki/Gimbal_lock#Loss_of_a_degree_of_freedom_with_Euler_angles
+     *
+     * The rotation matrix for y = +-90 degrees and x = 0 becomes
+     *
+     * Matrix3d.createRowValues(
+     *      0, -sz, +-cz,
+     *      0, cz, +-sz,
```

Comment:
Nit in the comment: to denote that m(0,2) and m(1,2) have the _opposite_ sign of m(2,0) = +-1 in the gimbal lock case, you can use `-+cz` and `-+sz` instead of `+-`.

Reply:
ok will fix